Feature/change lead edit form to lead input

AdityaKhatri opened the pull request:
## Changes

* Add bulk upload functionality
* Change old lead edit form to use graphql

Depends on: https://github.com/the-deep/server/pull/974

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [x] translations

AdityaKhatri committed:
Change lead edit form to graphql

AdityaKhatri committed:
Fix form in source details in tagging flow

AdityaKhatri committed:
Add error handling in lead save and update

AdityaKhatri committed:
Separate out use states from LeadEditForm

AdityaKhatri committed:
Add form in bulk upload

AdityaKhatri committed:
Make bulk upload form functional

AdityaKhatri committed:
Add bulk upload mutation

samshara commented on app/components/lead/LeadPreview/Preview/mimeTypes.ts:
```diff
@@ -29,11 +29,11 @@ export type DocsMimeType =
     | 'application/wps-office.docx';
 
-export type MimeTypes = HTMLMimeTypes | ImageMimeTypes | DocsMimeType;
+export type MimeTypes = string;
```
any reason we are making the type weak?

tnagorra replied:
We don't need a strong typecheck for MimeTypes.
User will not explicity define a mimetype and there isn't case where the typesafety is useful for mimetypes

samshara commented on app/components/selections/ProjectUserSelectInput/index.tsx:
```diff
@@ -0,0 +1,80 @@
+import useDebouncedValue from '#hooks/useDebouncedValue';
+
+const PROJECT_USERS = gql`
+    # query ProjectUser($search: String, $projectId: ID!) {
+    query ProjectUser($projectId: ID!) {
+        project(id: $projectId) {
+            # members(search: $search) {
+            members {
+                id
+                displayName
+            }
+        }
+    }
+`;
```
it would be better to remove these comments. we can add fixme.

tnagorra replied:
Does this comment work?

samshara commented on app/views/Project/EntryEdit/LeftPane/index.tsx:
```diff
@@ -260,7 +259,7 @@ function LeftPane(props: Props) {
                     </QuickActionDropdownMenu>
                     <QuickActionLink
-                        to={lead?.url ?? lead?.attachment?.file ?? ''}
+                        to={lead?.url ?? lead?.attachment?.file?.url ?? ''}
```
if we need to handle empty string, we must use || operator instead of ??

tnagorra replied:
Looks like  in this case, we don't want to pass undefined to "to" prop.
So, the current implementation looks good to me.

samshara commented on app/views/Project/Tagging/Sources/BulkUpload/FilesUploaded/index.tsx:
```diff
@@ -10,64 +11,145 @@ import {
+    const [
+        leadGroupOptions,
+        setLeadGroupOptions,
+    ] = useState<BasicLeadGroup[] | undefined | null>(undefined);
```
```suggestion
    ] = useState<BasicLeadGroup[] | undefined | null>();
```

samshara commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
+    const handleSubmit = useCallback(
+        () => {
+            if (!projectId) {
+                // eslint-disable-next-line no-console
+                console.error('No project id');
```
remove this

tnagorra commented on app/views/Project/Tagging/index.tsx:
```diff
@@ -179,14 +183,16 @@ function Tagging(props: Props) {
                 <LeadEditModal
-                    projectId={+project.id}
+                    projectId={project.id}
                     onClose={hideSingleSourceAddModal}
                     onLeadSaveSuccess={handleSingleLeadSaveSuccess}
                 />
             )}
-            {isBulkModalShown && (
+            {isBulkModalShown && project?.id && (
```
Either we should use `project?.id` or `project.id`

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/types.ts:
```diff
@@ -1,7 +1,16 @@
 
 export type SourceFileType = 'disk' | 'google-drive' | 'dropbox';
 
+export const sourceTypeMap: { [key in string]: LeadSourceTypeEnum } = {
+    disk: 'DISK',
+    'google-drive': 'GOOGLE_DRIVE',
+    dropbox: 'DROPBOX',
+};
```
Why not use `[key in SourceFileType]`?

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
+            footerActions={(
+                <Button
+                    name={undefined}
+                    disabled={formPristine}
```
This shoudl also be disabled when the form is being saved

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
+                formValidate,
+                setFormError,
+                (value) => {
+                    setLeadClientIds(value?.leads?.map((lead) => lead.clientId));
+                    const leads = (value.leads ?? []) as LeadInputType[];
+
+                    if (leads.length > 0) {
+                        bulkCreateLeads({
+                            variables: {
+                                projectId,
+                                leads,
+                            },
+                        });
+                    } else {
+                        alert.show(
+                            'Leads uploaded successfully!',
+                            { variant: 'success' },
+                        );
+                    }
```
Only save leads that have not errored

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
+        const selectedLeadObj = formValue?.leads?.find((lead) => lead.clientId === clientId);
+        setFormFieldValue((oldVal) => oldVal?.filter((lead) => lead.clientId !== clientId), 'leads');
+        if (selectedLeadObj?.attachment) {
+            setUploadedFiles((oldState: FileUploadResponse[]) => {
+                const updatedState = produce(oldState, (safeState) => {
+                    const index = safeState.findIndex(
+                        (file) => String(file.id) === selectedLeadObj.attachment,
+                    );
+                    if (index !== -1) {
+                        safeState.splice(index, 1);
+                    }
+                });
+                return updatedState;
```
Are we cleaning up the uploaded files from memory?
Is this also done after the leads are saved and all of the leads are cleared?

tnagorra replied:
Also, do we need to do this?

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
+        },
+    );
+
+    const handleLeadChange = useCallback(
+        (val: SetValueArg<PartialLeadType>, otherName: number | undefined) => {
+            onLeadChange(
+                (oldValue) => {
+                    const newVal = !isCallable(val)
+                        ? val
+                        : val(oldValue);
+                    return newVal;
+                },
+                otherName,
+            );
+        },
+        [onLeadChange],
+    );
```
Why are we defining handleLeadChange in this manner?

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
+                            { variant: 'success' },
+                        );
+                        setLeadsAdded(true);
+                        setFormFieldValue((oldValues) => (
+                            oldValues?.filter((lead) => !uploadedLeads.includes(lead.clientId))
+                        ), 'leads');
```
Don't we need to cleanup the attachments in memory?

tnagorra replied:
Also, this will clear out all of the leads that were errored (from the server or from the form)

tnagorra commented on app/views/Project/Tagging/Sources/BulkUpload/index.tsx:
```diff
@@ -1,65 +1,339 @@
 
     const [uploadedFiles, setUploadedFiles] = useState<FileUploadResponse[]>([]);
+    // NOTE: leadsAdded is a boolean set when at least on lead was successfully added
+    const [leadsAdded, setLeadsAdded] = useState<boolean>(false);
```
Let's add comment on why we have this boolean.

AdityaKhatri committed:
Fix remove comments

tnagorra committed:
Fix ProjectUserSelectInput api

tnagorra committed:
Identify "localhost" as valid link

- Show "localhost" links using iframe
- Remove padding from LeadPreview in EditEntry
- Remove assignee transform step for EntryEdit form init
- Fix error message on bulk upload error
- Show lead preview on lead add modal.
- Change lead preview when url is changed on lead edit modal.
  (even though we don't let users change url)
